### PR TITLE
Feature/rigol dg1062/adding duty cycle parameter

### PR DESCRIPTION
**Pull request by vhartong**
Changes proposed in this pull request:
- Added Pulse waveform to the list
- Added translation dictionary for waveforms (Device responses are sometimes different 👎 )
- Adding function duty cycle parameter for Square wave and Pulse Wave
- General code cleanup + two bugs

### CODE REVIEW

File: qcodes/instrument_drivers/rigol/DG1062.py [hunk 1]
```diff
@@ -160,8 +165,8 @@ def __init__(self, parent: 'DG1062', name: str, channel: int) ->None:
                 ),
                 vals.Enum("INF", "MIN", "MAX", "HighZ")
             ),
-            set_parser=lambda value: "INF" if value == "HighZ" else value,
             get_parser=lambda value: "HighZ"
+            set_parser=lambda value: "INF" if value == "HighZ" else value,
             if float(value) > DG1062Channel.max_impedance else float(value)
         )
```
Comment (astafan8) [inline qcodes/instrument_drivers/rigol/DG1062.py]: something terribly wrong happened here :) the automatic testing is alarming: https://ms-quantum.visualstudio.com/QubitWorkbench/_build/results?buildId=10600.

I think when moving `set_parser` inserted it in the middle of the lambda expression for `get_parser` which breaks everything :) I guess the following is what you wanted, right?:
```python
<...>
),  # <- comma from the line before
get_parser=lambda value: "HighZ" 
    if float(value) > DG1062Channel.max_impedance else float(value), # <- comma
set_parser=lambda value: "INF" if value == "HighZ" else value  # <- no comma here because of the brace on the next line
)
<...>
```

Also, you sure this final version of the code has been fully tested on a real device? :)

File: qcodes/instrument_drivers/rigol/DG1062.py [hunk 2]
```diff
@@ -102,20 +102,25 @@ def trigger(self) ->None:
...
-        ["HARM", "NOIS", "RAMP", "SIN", "SQU", "TRI", "USER"]
+        ["HARM", "NOIS", "RAMP", "SIN", "SQU", "TRI", "USER", "PULS"]
     }
 
     waveform_params["DC"] = ["freq", "ampl", "offset"]
     waveform_params["ARB"] = ["sample_rate", "ampl", "offset"]
 
+    """
+    Responses from the machine don't always match the name to set the function, hence a translater
+    """
```
Comment (astafan8) [inline qcodes/instrument_drivers/rigol/DG1062.py]: great catch!

File: qcodes/instrument_drivers/rigol/DG1062.py [hunk 3]
```diff
@@ -182,8 +187,15 @@ def __init__(self, parent: 'DG1062', name: str, channel: int) ->None:
 
         self.add_parameter(
             "state",
-            set_cmd=f"OUTPUT{channel}:STATE {{}}",
             get_cmd=f"OUTPUT{channel}:STATE?",
+            set_cmd=f"OUTPUT{channel}:STATE {{}}",
+        )
+        
+        self.add_parameter(
+            "dcycle",
```
Comment (astafan8) [inline qcodes/instrument_drivers/rigol/DG1062.py]: could you rename this to `duty_cycle` for readability? good names give much more long-term benefit that saving on a couple of characters (it does depend on a particular case, but here - for sure).

File: qcodes/instrument_drivers/rigol/DG1062.py [hunk 4]
```diff
@@ -182,8 +187,15 @@ def __init__(self, parent: 'DG1062', name: str, channel: int) ->None:
 
         self.add_parameter(
             "state",
-            set_cmd=f"OUTPUT{channel}:STATE {{}}",
             get_cmd=f"OUTPUT{channel}:STATE?",
+            set_cmd=f"OUTPUT{channel}:STATE {{}}",
+        )
+        
+        self.add_parameter(
```
Comment (astafan8) [inline qcodes/instrument_drivers/rigol/DG1062.py]: Would you mind adding a `label`, `unit`, and `docstring` to this parameter? (this is done by adding new arguments with these names to the `add_parameter` call, see how it's done in other drivers, even for some parameters in this driver)

File: qcodes/instrument_drivers/rigol/DG1062.py [hunk 5]
```diff
@@ -290,7 +304,32 @@ def _set_waveform_params(self, **params_dict: Dict) ->None:
         string += ",".join(
             ["{:7e}".format(params_dict[param]) for param in param_names])
         self.parent.write_raw(string)
+        
+    def _get_duty_cycle(self):
+        """
+        Simple function to extract the duty cycle after checking waveform
+        """
+        self.waveform()
+        wf = self.waveform.get_latest()
```
Comment (astafan8) [inline qcodes/instrument_drivers/rigol/DG1062.py]: in case it's necessary for `_get_duty_cycle` function to operate on the real actual latest waveform of the instrument, then `wf = self.waveform.get_latest()` is redundant because `wf = self.waveform()` does it already.

in case it's necessary for `_get_duty_cycle` function to avoid additional communication with the instrument, then `self.waveform()` should be removed because it does perform the communication; while `wf = self.waveform.get_latest()` only does the job of returning the "latest value of the parameter that is known to the driver (yes, the state of the instrument might have changed since the value of the parameter has been requested due to, for example, manually clicking buttons of the instrument)".

So, depending on the intention, please, get rid of either of the lines.

File: qcodes/instrument_drivers/rigol/DG1062.py [hunk 6]
```diff
@@ -290,7 +304,32 @@ def _set_waveform_params(self, **params_dict: Dict) ->None:
...
+
+        if wf == 'PULS' or wf == 'SQU':
```
Comment (astafan8) [inline qcodes/instrument_drivers/rigol/DG1062.py]: the following is more pythonic, readable, future-proof
```suggestion
        if wf in ['PULS', 'SQU']:
```

File: qcodes/instrument_drivers/rigol/DG1062.py [hunk 7]
```diff
@@ -290,7 +304,32 @@ def _set_waveform_params(self, **params_dict: Dict) ->None:
...
+            duty_cycle = self.parent.ask_raw(f":SOUR{self.channel}:FUNC:{wf}:DCYC?")
```
Comment (astafan8) [inline qcodes/instrument_drivers/rigol/DG1062.py]: It is quite weird that throughout this driver `self.parent.ask_raw` and `self.parent.write_raw` are used instead of just `self.ask` and `self.write` which do the same, but are much cleaner to use from the code/design point of view.

For this driver and this PR, ok, let's keep it consistent. But in general, it's expected that `self.write` and `self.ask` of the `InstrumentChannel` (which is the class of `self` here) class are used, unless there is some really important reason not to do it.

File: qcodes/instrument_drivers/rigol/DG1062.py [hunk 8]
```diff
@@ -290,7 +304,32 @@ def _set_waveform_params(self, **params_dict: Dict) ->None:
         string += ",".join(
             ["{:7e}".format(params_dict[param]) for param in param_names])
         self.parent.write_raw(string)
+        
+    def _get_duty_cycle(self):
+        """
+        Simple function to extract the duty cycle after checking waveform
```
Comment (astafan8) [inline qcodes/instrument_drivers/rigol/DG1062.py]: This docstring has ~70% of useful information - "extract the duty cycle after checking waveform"  - and ~30% of useless information - "Simple function to". Could you rephrase it to smth like "Read duty cycle value. Also reads the waveform parameter of the instrument."?

Moreover, one might argue that the fact that this method raises when used for waveforms that don't have a meaningful duty cycle, should also be mentioned in the docstring. In this particular case, this note belongs to the docstring of the parameter (see the `docstring` kwargs for `self.add_parameter('duty_cycle', ...)`.

Finally, this method is fairly short, hence it's code perfectly expresses what is going on, hence the docstring is not really needed, in my humble opinion.

To conclude, please:
* depending on your taste, either remove the docstring or make it "100%" useful :)
* add a notion about "allowed waveforms" to the docstring of the parameter in `add_parameter(...)` call.

File: qcodes/instrument_drivers/rigol/DG1062.py [hunk 9]
```diff
@@ -290,7 +304,32 @@ def _set_waveform_params(self, **params_dict: Dict) ->None:
...
+        wf = self.waveform.get_latest()
+
+        if wf == 'PULS' or wf == 'SQU':
+            duty_cycle = self.parent.ask_raw(f":SOUR{self.channel}:FUNC:{wf}:DCYC?")
+        else:
+            raise ValueError(f"Can not read duty cycle for current function: {wf}")
 
+        return duty_cycle
+
+    def _set_duty_cycle(self,duty_cycle):
```
Comment (astafan8) [inline qcodes/instrument_drivers/rigol/DG1062.py]: The same comments from `_get_duty_cycle` apply to this method.

File: qcodes/instrument_drivers/rigol/DG1062.py [hunk 10]
```diff
@@ -290,7 +304,32 @@ def _set_waveform_params(self, **params_dict: Dict) ->None:
         string += ",".join(
             ["{:7e}".format(params_dict[param]) for param in param_names])
         self.parent.write_raw(string)
+        
+    def _get_duty_cycle(self):
```
Comment (astafan8) [inline qcodes/instrument_drivers/rigol/DG1062.py]: what's the type of the return value?`float`? if so, please, add a type annotation like this:
```suggestion
    def _get_duty_cycle(self) -> float:
```

File: qcodes/instrument_drivers/rigol/DG1062.py (hunk 1 above)
Comment (vhartong) [reply qcodes/instrument_drivers/rigol/DG1062.py]: Yes, this caused a problem. Thanks for picking it up

File: qcodes/instrument_drivers/rigol/DG1062.py (hunk 3 above)
Comment (vhartong) [reply qcodes/instrument_drivers/rigol/DG1062.py]: Done.

File: qcodes/instrument_drivers/rigol/DG1062.py (hunk 4 above)
Comment (vhartong) [reply qcodes/instrument_drivers/rigol/DG1062.py]: duty cycle does not have a unit, writing up doc_string: done

File: qcodes/instrument_drivers/rigol/DG1062.py (hunk 6 above)
Comment (vhartong) [reply qcodes/instrument_drivers/rigol/DG1062.py]: completely agree, looks way better

File: qcodes/instrument_drivers/rigol/DG1062.py (hunk 7 above)
Comment (vhartong) [reply qcodes/instrument_drivers/rigol/DG1062.py]: to be honest, i tried to stay consistent within the code. maybe something for the next merge?

Comment (astafan8) [reply qcodes/instrument_drivers/rigol/DG1062.py]: sure.
